tests/unit/metrics/test_ffwd.py

```diff
@@ -0,0 +1,208 @@
+    assert ('1.2.3.4', '5678') == actual_addr
+    assert b'{"an": "object"}' == actual_protocol.message
+
+
+@pytest.fixture
+def mock_udp_client_and_metrics_sent(mocker):
+    metrics_sent = []
+
+    async def mock_send(metric):
+        nonlocal metrics_sent
+        metrics_sent.append(metric)
+
+    mock_udp_client = mocker.Mock()
+    mock_udp_client.send = mock_send
+    return mock_udp_client, metrics_sent
+
+
+@pytest.fixture
+def timer_and_metrics_sent(mock_udp_client_and_metrics_sent):
+    mock_udp_client, metrics_sent = mock_udp_client_and_metrics_sent
+    metric = {'a': 'metric', 'value': None}
+    timer = ffwd.FfwdTimer(metric, mock_udp_client)
+    return timer, metrics_sent
+
+
+@pytest.mark.asyncio
+async def test_ffwdtimer_manual(timer_and_metrics_sent):
+    """Times when operated manually."""
+    timer, metrics_sent = timer_and_metrics_sent
+    await timer.start()
+    assert isinstance(timer._start_time, numbers.Number)
+
+    await timer.stop()
+    assert len(metrics_sent) == 1
+    assert 'metric' == metrics_sent[0]['a']
+    assert 0 < metrics_sent[0]['value']
+
+
+@pytest.mark.asyncio
+async def test_ffwdtimer_as_context_manager(timer_and_metrics_sent):
+    """Times as a context manager."""
+    timer, metrics_sent = timer_and_metrics_sent
+    async with timer as t:
+        assert isinstance(t._start_time, numbers.Number)
+
+    assert len(metrics_sent) == 1
+    assert 'metric' == metrics_sent[0]['a']
+    assert 0 < metrics_sent[0]['value']
+
+
+@pytest.fixture
+def test_config(mocker, monkeypatch):
+    monkeypatch.setattr(ffwd, 'UDPClient', mocker.Mock)
+    return {
+        'key': 'service-name',
+        'ffwd_ip': '1.2.3.4',
+        'ffwd_port': '5678',
+        'time_unit': 1E9
+    }
+
+
+@pytest.mark.parametrize('key,name,val,attrs,tags', [
+    ('service-name', 'metric-name', 10, None, None),
+    ('other-name', 'other-metric', 6.243289, {'hostname': 'dns-host-3'}, None),
+    ('service-name', 'other-metric', 500000, None, ['v40']),
+    ('service-name', 'other-metric', 500000, {'type': 'active'}, ['v40'])
+])
+def test_ffwdrelay_create_metric(test_config, key, name, val, attrs, tags):
+    """Creates metric for all arguments."""
+    test_config['key'] = key
+    relay = ffwd.FfwdRelay(test_config)
+    actual = relay._create_metric(name, val, attrs=attrs, tags=tags)
+
+    attrs = attrs or {}
+    attrs.update({'what': name})
+    expected = {
+        'key': key,
+        'attributes': attrs,
+        'value': val,
+        'type': 'metric',
+        'tags': tags or []
+    }
+    assert expected == actual
+
+
+@pytest.fixture
+def ffwdrelay_and_metrics_sent(test_config, mock_udp_client_and_metrics_sent):
+    mock_udp_client, metrics_sent = mock_udp_client_and_metrics_sent
+    relay = ffwd.FfwdRelay(test_config)
+    relay.udp_client = mock_udp_client
+    return relay, metrics_sent
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize('attrs,tags', [
```

Comment:
Maybe pull this array out instead of repeating it?

Reply:
Fair enough.